internal/storage/v2/memory/tenant.go

```
@@ -123,6 +136,73 @@ func (t *Tenant) getTraces(traceIds ...tracestore.GetTraceParams) []ptrace.Trace
+		return nil, errors.New("start time is required")
+	}
+	if !query.EndTime.IsZero() && query.EndTime.Before(query.StartTime) {
+		return nil, errors.New("end time must be greater than start time")
+	}
+	t.RLock()
+	defer t.RUnlock()
+	deps := map[string]*model.DependencyLink{}
+	for _, index := range t.ids {
+		traceWithTime := t.traces[index]
+		if traceIsBetweenStartAndEnd(traceWithTime, query.StartTime, query.EndTime) {
+			for _, resourceSpan := range traceWithTime.trace.ResourceSpans().All() {
+				for _, scopeSpan := range resourceSpan.ScopeSpans().All() {
+					for _, span := range scopeSpan.Spans().All() {
+						if !span.ParentSpanID().IsEmpty() {
+							spanServiceName := getServiceNameFromResource(resourceSpan.Resource())
+							parentSpanServiceName, found := findServiceNameWithSpanId(traceWithTime.trace, span.ParentSpanID())
+							if found {
+								depKey := parentSpanServiceName + "&&&" + spanServiceName
+								if _, ok := deps[depKey]; !ok {
+									deps[depKey] = &model.DependencyLink{
+										Parent:    parentSpanServiceName,
+										Child:     spanServiceName,
+										CallCount: 1,
+									}
+								} else {
+									deps[depKey].CallCount++
+								}
+							}
+						}
+					}
+				}
+			}
+		}
+	}
+	retMe := make([]model.DependencyLink, 0, len(deps))
+	for _, dep := range deps {
+		retMe = append(retMe, *dep)
+	}
+	return retMe, nil
+}
+
+func findServiceNameWithSpanId(trace ptrace.Traces, spanId pcommon.SpanID) (string, bool) {
+	for _, resourceSpan := range trace.ResourceSpans().All() {
+		for _, scopeSpan := range resourceSpan.ScopeSpans().All() {
+			for _, span := range scopeSpan.Spans().All() {
+				if span.SpanID() == spanId {
+					return getServiceNameFromResource(resourceSpan.Resource()), true
+				}
+			}
+		}
+	}
+	return "", false
+}
+
+func traceIsBetweenStartAndEnd(traceWithTime traceAndId, startTime time.Time, endTime time.Time) bool {
+	if endTime.IsZero() && traceWithTime.startTime.After(startTime) {
```

Rewrite to fail fast and you won't need to repeat calculations